### ISSUE TITLE
URL parts mapping to $parameters

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#267]
**Reported by @stefanb on 30 Aug 2003 13:26 UTC**
Inspired by bug 143 and bug 412 i came to an idea, that "ugly" location
http://server/context/path/to/some.jsp?bla=foo&bar=42
could be reached via nicer URL
http://server/context/path/to/some/foo/42
if the controller knew how to transform certain parts of url as new request
$attributes. In this case it would need to know the pattern:
"path/to/some/$foo/$bar"

Ideally, the mechanism would be so flexible that varieties like:
http://server/context/path/to/42/foo
or http://server/context/foo/something/42
would be also supported. 
(not necesarily many of them at the same time as this would lead to inconsistent
URL structure)

This could be specified in the logic classes as field:
public String attributePattern="/path/to/some/$foo/$bar";
and whenever $attribute (eg $foo) would be required (eg in OQL) it would be
extracted from the URL (or controller could pass on the request to .jsp with
these additional parameters).

EventsEventLogic could map requests
/events/event/19-G-LJ         -> /events/event.jsp?event=19-G-LJ
with attributePattern="/events/event/$event";

EventsapplicantListLogic could map requests
/events/applicantList/19-G-LJ -> /events/applicantList.jsp?event=19-G-LJ
with attributePattern="/events/applicantList/$event";

Probably jsp base filename (some,event,applicantList) could be omitted from the
URL altogether and make it possible to map requests:
/events/19-G-LJ            ->/events/event.jsp?event=19-G-LJ
/events/19-G-LJ/applicants ->/events/applicantList.jsp?event=19-G-LJ
/events/19-G-LJ/apply      ->/events/applicant/applicationNew.jsp?event=19-G-LJ
with some field in EventLogic mapping URL patterns
"/events/$event", "/events/$event/applicants" and "/events/$event/apply"
to their JSPs.

This brings me to the idea that maybe it would be more convinient to specify
these mappings within the requested JSP itself. Eg in above mentioned some.jsp
as: <%!public String attributeURLPattern="/path/to/some/$foo/$bar";%>
or <%!public String attributeURLPattern="/$foo/something/$bar";%>

but then the mappings from URL to specific jsp needs to be done somewhere else
(currently this is possible in web.xml), or makumba controller needs to look in
all jsps for their public URL pattern.

I believe there is plenty of room to improve this initial design :)
The real test of the design would also be to provide Kramba's
"/archive/doc/$docNr/$attachmentName" functionality easily with it.

Basically this would gives some power of Apache's Mod_Rewrite
(http://httpd.apache.org/docs/mod/mod_rewrite.html ) to makumba aplication
programmers, mostly what they seem to most commonly need.

Maybe there are existing filters that can be used in tomcat for this purpose.

Migrated-From: http://trac.makumba.org/ticket/547

**Comment by ghost** [comment]
**Comment by @frederikhabils on 31 Aug 2003 22:37 UTC**
interesting idea. 

one thing i want to mention as a 'risk' to the overly enthusiastic use of these
fancier url's. 

When page X is accessible 
as      http://server/dir1/dir2/page.jsp?par=X&par2=Y
And as  http://server/dir1/dir2/page/X/Y

then you'll need absolute links in the page, because you risk to break relative
links. 
Suppose this page links to <a href=../season/view.jsp>

if browser has loaded http://server/dir1/dir2/page.jsp?par=X&par2=Y
it will request       http://server/dir1/season/view.jsp

if browser has loaded http://server/dir1/dir2/page/X/Y
it will request       http://server/dir1/dir2/page/season/view.jsp

which can be a mess.


I'm also not sure how you link this to bug 412, which states that having a
dynamic part (ie. a querystring) in the URL, will help avoid improper caching.
This fancy URL will actually hide the querystring from the browser/network, and
increase improper caching. (note: improper caching might also be resolved with
anti-caching settings on the http response, etc, etc.)

Finally, I do like this idea of easing the construction of fancy URL, where
their use is following from good design thinking. So let's see how to do that :-)

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 1 Sep 2003 00:00 UTC**
i think it's pretty hard to standardise this, as there are so many ways of doing it

on the other hand, i think it should be possible to write a servlet (makumba
application or separate) that can rewrite URLs in the ways described (which
would then replace the code using for archives now, and move it from JSP to Java)

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
